File: client/src/main/java/au/com/codeka/warworlds/client/concurrency/ThreadPool.java
package au.com.codeka.warworlds.client.concurrency;

import androidx.annotation.NonNull;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Executor;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * A pool of threads that we use for running things on {@link Threads#BACKGROUND}.
 */
public class ThreadPool {
  private final Threads thread;
  private final Executor executor;

  /**
   * Constructs a new {@link ThreadPool}.
   * @param thread The {@link Threads} of the thread pool, that we use to name individual threads.
   * @param maxQueuedItems The maximum number of items we'll allow to be queued.
   * @param minThreads The minimum number of threads in the thread pool.
   * @param maxThreads The maximum number of threads in the thread pool.
   * @param keepAliveMs The number of milliseconds to keep an idle thread in the thread pool.
   */
  public ThreadPool(
      final Threads thread,
      int maxQueuedItems,
      int minThreads,
      int maxThreads,
      int keepAliveMs) {
    this.thread = thread;

    ThreadFactory threadFactory = new ThreadFactory() {
      private final AtomicInteger count = new AtomicInteger(1);
      public Thread newThread(@NonNull Runnable r) {
        return new Thread(r, thread.toString() + " #" + count.getAndIncrement());
      }
    };

    BlockingQueue<Runnable> workQueue = new LinkedBlockingQueue<>(maxQueuedItems);

    executor = new ThreadPoolExecutor(
        minThreads, maxThreads, keepAliveMs, TimeUnit.MILLISECONDS, workQueue, threadFactory);
  }

  public void run(Runnable runnable) {
    executor.execute(runnable);
  }

  public boolean isThread(Threads thread) {
    return thread == this.thread;
  }
}

File: client/src/main/kotlin/au/com/codeka/warworlds/client/concurrency/ThreadPool.kt
package au.com.codeka.warworlds.client.concurrency

import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicInteger

/** A pool of threads that we use for running things on [Threads.BACKGROUND]. */
class ThreadPool(
    private val thread: Threads,
    maxQueuedItems: Int,
    minThreads: Int,
    maxThreads: Int,
    keepAliveMs: Long) {
  val executor: ThreadPoolExecutor
  val openedThreads: HashSet<Thread> = HashSet()

  fun run(runnable: Runnable) {
    executor.execute(runnable)
  }

  fun isThread(thread: Threads): Boolean {
    return thread == this.thread
  }

  /** Returns true if the given thread is part of our thread pool. */
  fun ownsThread(thread: Thread): Boolean {
    synchronized(openedThreads) {
      return openedThreads.contains(thread)
    }
  }

  /**
   * Constructs a new [ThreadPool].
   * @param thread The [Threads] of the thread pool, that we use to name individual threads.
   * @param maxQueuedItems The maximum number of items we'll allow to be queued.
   * @param minThreads The minimum number of threads in the thread pool.
   * @param maxThreads The maximum number of threads in the thread pool.
   * @param keepAliveMs The number of milliseconds to keep an idle thread in the thread pool.
   */
  init {
    val threadFactory: ThreadFactory = object : ThreadFactory {
      private val count = AtomicInteger(1)

      override fun newThread(r: Runnable): Thread {
        // Delete any threads from openedThreads that are done.
        synchronized(openedThreads) {
          val toRemove = ArrayList<Thread>()
          for (thread in openedThreads) {
            if (!thread.isAlive) {
              toRemove.add(thread)
            }
          }
          for (thread in toRemove) {
            openedThreads.remove(thread)
          }

          val thread = Thread(r, thread.toString() + " #" + count.getAndIncrement())
          openedThreads.add(thread)
          return thread
        }
      }
    }

    val workQueue: BlockingQueue<Runnable> = LinkedBlockingQueue(maxQueuedItems)

    executor = ThreadPoolExecutor(
        minThreads, maxThreads, keepAliveMs, TimeUnit.MILLISECONDS, workQueue, threadFactory)
  }

}